processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferArrayGrouper.java

```diff
@@ -182,29 +186,31 @@ private void initializeSlot(int dimIndex)
 
   private boolean isUsedSlot(int dimIndex)
   {
-    final int index = dimIndex / 8;
-    final int extraIndex = dimIndex % 8;
-    final int usedByte = 1 << extraIndex;
-    return (usedFlagBuffer.get(index) & usedByte) == usedByte;
+    final int index = dimIndex / Byte.SIZE;
+    final int extraIndex = dimIndex % Byte.SIZE;
+    final int usedFlagByte = 1 << extraIndex;
+    return (usedFlagBuffer.get(index) & usedFlagByte) != 0;
   }
 
   @Override
   public void reset()
   {
-    final int usedBufferCapacity = usedFlagBuffer.capacity();
+    // Clear the entire usedFlagBuffer
+    final int usedFlagBufferCapacity = usedFlagBuffer.capacity();
 
-    final int n = usedBufferCapacity / 8;
-    for (int i = 0; i < n; i += 8) {
+    // putLong() instead of put() can boost the performance of clearing the buffer
+    final int n = usedFlagBufferCapacity / Long.BYTES;
+    for (int i = 0; i < n; i += Long.BYTES) {
```

Comment:
Seems that there is a bug in this logic, it will clear using putLong() only 1/8-th of the buffer capacity, instead of the buffer capacity without, at most, 7 bytes. The classic pattern of implementing such things is
```java
int i = 0;
for (; i <= usedFlagBufferCapacity - Long.BYTES; i += Long.BYTES) {
  usedFlagBuffer.putLong(i, 0L);
}
for (; i < usedFlagBufferCapacity; i++) {
  usedFlagBuffer.put(i, (byte) 0);
}